File: apps/dashboard/src/features/careers/schemas.ts
import { z } from "zod";

import {
  applicationSources,
  applicationStatuses,
  interviewStatuses,
  interviewTypes,
  jobListingStatuses,
} from "./types";

const nullableDecimalSchema = z
  .union([z.string(), z.number()])
  .nullable()
  .transform((value) => (value === null ? null : String(value)));

export const careersDashboardSchema = z.object({
  open_listings: z.number(),
  featured_listings: z.number(),
  total_applications: z.number(),
  new_applications: z.number(),
  screening_applications: z.number(),
  shortlisted_applications: z.number(),
  interview_applications: z.number(),
  offered_applications: z.number(),
  hired_applications: z.number(),
  rejected_applications: z.number(),
  upcoming_interviews: z.number(),
  completed_interviews: z.number(),
  applications_by_status: z.record(z.string(), z.number()),
  applications_by_source: z.record(z.string(), z.number()),
});

export const jobDepartmentSchema = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
  description: z.string(),
  is_active: z.boolean(),
  sort_order: z.number(),
});

export const employmentTypeSchema = z.object({
  id: z.string(),
  name: z.string(),
  code: z.string(),
  description: z.string(),
  is_active: z.boolean(),
  sort_order: z.number(),
});

export const jobPositionSchema = z.object({
  id: z.string(),
  department_id: z.string(),
  department_name: z.string(),
  employment_type_id: z.string(),
  employment_type_name: z.string(),
  title: z.string(),
  slug: z.string(),
  summary: z.string(),
  description: z.record(z.string(), z.unknown()),
  responsibilities: z.array(z.unknown()),
  requirements: z.array(z.unknown()),
  preferred_qualifications: z.array(z.unknown()),
  benefits: z.array(z.unknown()),
  location: z.string(),
  remote_policy: z.string(),
  experience_level: z.string(),
  salary_min: nullableDecimalSchema,
  salary_max: nullableDecimalSchema,
  salary_currency: z.string(),
  salary_visible: z.boolean(),
  is_active: z.boolean(),
  sort_order: z.number(),
});

export const jobListingSchema = z.object({
  id: z.string(),
  position_id: z.string(),
  position_title: z.string(),
  department_name: z.string(),
  employment_type_name: z.string(),
  reference_code: z.string(),
  status: z.enum(jobListingStatuses),
  number_of_openings: z.number(),
  application_deadline: z.string().nullable(),
  published_at: z.string().nullable(),
  scheduled_for: z.string().nullable(),
  is_featured: z.boolean(),
  is_active: z.boolean(),
  is_publicly_available: z.boolean(),
  sort_order: z.number(),
  created_at: z.string(),
  updated_at: z.string(),
});

const applicationAnswerSchema = z.object({
  id: z.string(),
  question_id: z.string(),
  question: z.string(),
  answer: z.unknown(),
});

export const applicationNoteSchema = z.object({
  id: z.string(),
  author_id: z.string().nullable(),
  author_name: z.string().nullable(),
  note: z.string(),
  is_private: z.boolean(),
  created_at: z.string(),
});

export const jobApplicationSchema = z.object({
  id: z.string(),
  listing_id: z.string(),
  listing_reference_code: z.string(),
  position_title: z.string(),
  applicant_name: z.string(),
  email: z.string(),
  phone: z.string(),
  country: z.string(),
  city: z.string(),
  linkedin_url: z.string(),
  portfolio_url: z.string(),
  current_company: z.string(),
  current_position: z.string(),
  years_of_experience: nullableDecimalSchema,
  expected_salary: nullableDecimalSchema,
  expected_salary_currency: z.string(),
  availability_date: z.string().nullable(),
  cover_letter: z.string(),
  resume_asset_id: z.string().nullable(),
  status: z.enum(applicationStatuses),
  source: z.enum(applicationSources),
  assigned_to_id: z.string().nullable(),
  assigned_to_name: z.string().nullable(),
  submitted_at: z.string(),
  reviewed_at: z.string().nullable(),
  rating: z.number().nullable(),
  internal_summary: z.string(),
  rejection_reason: z.string(),
  consent_to_process: z.boolean(),
  consent_to_retain: z.boolean(),
  answers: z.array(applicationAnswerSchema),
  notes: z.array(applicationNoteSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

const interviewParticipantSchema = z.object({
  id: z.string(),
  user_id: z.string(),
  user_name: z.string(),
  is_lead: z.boolean(),
  attendance_confirmed: z.boolean(),
});

export const interviewSchema = z.object({
  id: z.string(),
  application_id: z.string(),
  applicant_name: z.string(),
  position_title: z.string(),
  title: z.string(),
  interview_type: z.enum(interviewTypes),
  status: z.enum(interviewStatuses),
  scheduled_start: z.string(),
  scheduled_end: z.string(),
  timezone_name: z.string(),
  location: z.string(),
  meeting_url: z.string(),
  instructions: z.string(),
  organizer_id: z.string().nullable(),
  organizer_name: z.string().nullable(),
  completed_at: z.string().nullable(),
  cancellation_reason: z.string(),
  participants: z.array(interviewParticipantSchema),
  created_at: z.string(),
  updated_at: z.string(),
});

File: backend/apps/careers/schemas.py
from datetime import datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from ninja import Schema
from pydantic import Field


class JobDepartmentSchema(Schema):
    id: UUID
    name: str
    slug: str
    description: str
    is_active: bool
    sort_order: int


class EmploymentTypeSchema(Schema):
    id: UUID
    name: str
    code: str
    description: str
    is_active: bool
    sort_order: int


class JobPositionSchema(Schema):
    id: UUID
    department_id: UUID
    department_name: str
    employment_type_id: UUID
    employment_type_name: str
    title: str
    slug: str
    summary: str
    description: dict
    responsibilities: list
    requirements: list
    preferred_qualifications: list
    benefits: list
    location: str
    remote_policy: str
    experience_level: str
    salary_min: Decimal | None
    salary_max: Decimal | None
    salary_currency: str
    salary_visible: bool
    is_active: bool
    sort_order: int


class JobListingSchema(Schema):
    id: UUID
    position_id: UUID
    position_title: str
    department_name: str
    employment_type_name: str
    reference_code: str
    status: str
    number_of_openings: int
    application_deadline: datetime | None
    published_at: datetime | None
    scheduled_for: datetime | None
    is_featured: bool
    is_active: bool
    is_publicly_available: bool
    sort_order: int
    created_at: datetime
    updated_at: datetime



ApplicationStatusValue = Literal[
    "new",
    "screening",
    "shortlisted",
    "interview",
    "assessment",
    "offered",
    "hired",
    "rejected",
    "withdrawn",
    "archived",
]

ApplicationSourceValue = Literal[
    "careers_page",
    "linkedin",
    "facebook",
    "referral",
    "email",
    "manual",
    "other",
]

QuestionTypeValue = Literal[
    "short_text",
    "long_text",
    "number",
    "boolean",
    "single_choice",
    "multiple_choice",
    "date",
]


class ApplicationQuestionCreateSchema(Schema):
    question: str
    question_type: QuestionTypeValue = "short_text"
    help_text: str = ""
    options: list = Field(default_factory=list)
    is_required: bool = False
    is_active: bool = True
    sort_order: int = 0


class ApplicationQuestionSchema(Schema):
    id: UUID
    listing_id: UUID
    question: str
    question_type: str
    help_text: str
    options: list
    is_required: bool
    is_active: bool
    sort_order: int


class ApplicationAnswerInputSchema(Schema):
    question_id: UUID
    answer: object


class JobApplicationCreateSchema(Schema):
    listing_id: UUID
    applicant_name: str
    email: str
    phone: str = ""
    country: str = ""
    city: str = ""
    linkedin_url: str = ""
    portfolio_url: str = ""
    current_company: str = ""
    current_position: str = ""
    years_of_experience: Decimal | None = None
    expected_salary: Decimal | None = None
    expected_salary_currency: str = "LKR"
    availability_date: str | None = None
    cover_letter: str = ""
    resume_asset_id: UUID | None = None
    source: ApplicationSourceValue = "manual"
    assigned_to_id: UUID | None = None
    consent_to_process: bool = False
    consent_to_retain: bool = False
    answers: list[ApplicationAnswerInputSchema] = Field(default_factory=list)


class JobApplicationStatusSchema(Schema):
    status: ApplicationStatusValue
    rejection_reason: str = ""


class ApplicationNoteCreateSchema(Schema):
    note: str
    is_private: bool = True


class ApplicationAnswerSchema(Schema):
    id: UUID
    question_id: UUID
    question: str
    answer: object


class ApplicationNoteSchema(Schema):
    id: UUID
    author_id: UUID | None
    author_name: str | None
    note: str
    is_private: bool
    created_at: datetime


class JobApplicationSchema(Schema):
    id: UUID
    listing_id: UUID
    listing_reference_code: str
    position_title: str
    applicant_name: str
    email: str
    phone: str
    country: str
    city: str
    linkedin_url: str
    portfolio_url: str
    current_company: str
    current_position: str
    years_of_experience: Decimal | None
    expected_salary: Decimal | None
    expected_salary_currency: str
    availability_date: str | None
    cover_letter: str
    resume_asset_id: UUID | None
    status: str
    source: str
    assigned_to_id: UUID | None
    assigned_to_name: str | None
    submitted_at: datetime
    reviewed_at: datetime | None
    rating: int | None
    internal_summary: str
    rejection_reason: str
    consent_to_process: bool
    consent_to_retain: bool
    answers: list[ApplicationAnswerSchema]
    notes: list[ApplicationNoteSchema]
    created_at: datetime
    updated_at: datetime



InterviewStatusValue = Literal[
    "scheduled",
    "confirmed",
    "completed",
    "cancelled",
    "no_show",
    "rescheduled",
]

InterviewTypeValue = Literal[
    "phone",
    "video",
    "onsite",
    "technical",
    "hr",
    "final",
]

EvaluationRecommendationValue = Literal[
    "strong_hire",
    "hire",
    "hold",
    "no_hire",
    "strong_no_hire",
]


class InterviewParticipantInputSchema(Schema):
    user_id: UUID
    is_lead: bool = False


class InterviewCreateSchema(Schema):
    application_id: UUID
    title: str = "Job Interview"
    interview_type: InterviewTypeValue = "video"
    scheduled_start: datetime
    scheduled_end: datetime
    timezone_name: str = "Asia/Colombo"
    location: str = ""
    meeting_url: str = ""
    instructions: str = ""
    organizer_id: UUID | None = None
    participants: list[
        InterviewParticipantInputSchema
    ] = Field(default_factory=list)


class InterviewStatusSchema(Schema):
    status: InterviewStatusValue
    cancellation_reason: str = ""


class InterviewParticipantSchema(Schema):
    id: UUID
    user_id: UUID
    user_name: str
    is_lead: bool
    attendance_confirmed: bool


class InterviewSchema(Schema):
    id: UUID
    application_id: UUID
    applicant_name: str
    position_title: str
    title: str
    interview_type: str
    status: str
    scheduled_start: datetime
    scheduled_end: datetime
    timezone_name: str
    location: str
    meeting_url: str
    instructions: str
    organizer_id: UUID | None
    organizer_name: str | None
    completed_at: datetime | None
    cancellation_reason: str
    participants: list[
        InterviewParticipantSchema
    ]
    created_at: datetime
    updated_at: datetime


class ApplicationEvaluationCreateSchema(Schema):
    application_id: UUID
    interview_id: UUID | None = None
    evaluator_id: UUID | None = None
    technical_score: int | None = None
    communication_score: int | None = None
    culture_score: int | None = None
    overall_score: int
    recommendation: EvaluationRecommendationValue
    strengths: str = ""
    concerns: str = ""
    comments: str = ""


class ApplicationEvaluationSchema(Schema):
    id: UUID
    application_id: UUID
    applicant_name: str
    interview_id: UUID | None
    evaluator_id: UUID
    evaluator_name: str
    technical_score: int | None
    communication_score: int | None
    culture_score: int | None
    overall_score: int
    recommendation: str
    strengths: str
    concerns: str
    comments: str
    submitted_at: datetime
    created_at: datetime
    updated_at: datetime



class JobListingScheduleSchema(Schema):
    scheduled_for: datetime


class JobApplicationReviewSchema(Schema):
    assigned_to_id: UUID | None = None
    rating: int | None = None
    internal_summary: str = ""


class CareersDashboardSchema(Schema):
    open_listings: int
    featured_listings: int
    total_applications: int
    new_applications: int
    screening_applications: int
    shortlisted_applications: int
    interview_applications: int
    offered_applications: int
    hired_applications: int
    rejected_applications: int
    upcoming_interviews: int
    completed_interviews: int
    applications_by_status: dict
    applications_by_source: dict
